### ISSUE TITLE
descriptors evaluation  result

### ISSUE THREAD

**Issue by yanhuanhuanyy** [rpautrat/SuperPoint#93]

**Comment by yanhuanhuanyy** [comment]
@rpautrat ,  I am sorry for misoperation and the lack of a description. 

 I want to achieve descriptor evaluation results like yours, but failed. 
Steps I followed ( like #74 ) ,  I used default config parameters in all runs:
1. Training MagicPoint on Synthetic Shapes (40 000 iterations)
https://github.com/yanhuanhuanyy/superpoint_logfile/blob/master/magic-point_synth-log
2 and 3. Exporting detections on MS-COCO and Training MagicPoint on MS-COCO
command:
python export_detections.py configs/magic-point_coco_export.yaml magic-point_synth --pred_only --batch_size=5 --export_name=magic-point_coco-export1
# Change the parameter 'labels' in magic-point_coco_train.yaml to the folder magic-point_coco-export1
python experiment.py train configs/magic-point_coco_train.yaml magic-point_coco
python export_detections.py configs/magic-point_coco_export.yaml magic-point_coco --pred_only --batch_size=5 --export_name=magic-point_coco-export2
# Change the parameter 'labels' in magic-point_coco_train.yaml to the folder magic-point_coco-export2
python experiment.py train configs/magic-point_coco_train.yaml magic-point_coco
python export_detections.py configs/magic-point_coco_export.yaml magic-point_coco --pred_only --batch_size=5 --export_name=magic-point_coco-export3
https://github.com/yanhuanhuanyy/superpoint_logfile/blob/master/magic-point_coco-log
4. Training of SuperPoint on MS-COCO
"To avoid the OOM, I changed both batch_size and eval_batch_size to 1" 
command:
# Change the parameter 'labels' in superpoint_coco.yaml to the folder magic-point_coco-export3
python experiment.py train configs/superpoint_coco.yaml superpoint_coco
https://github.com/yanhuanhuanyy/superpoint_logfile/blob/master/superpoint_coco-log
5. Evaluating the repeatability on HPatches (detector_repeatability_hpatches.ipynb)
**HPatches Viewpoint**                                               
    magic-point_hpatches-repeat-v: 0.387                      
    magicLeap_mp_hpatches-repeat-v: 0.379                 
    harris_hpatches-repeat-v: 0.474                                
    fast_hpatches-repeat-v: 0.407                                    
    shi_hpatches-repeat-v: 0.407                                     
**HPatches Illumination**
    magic-point_hpatches-repeat-i: 0.678
    magicLeap_mp_hpatches-repeat-i: 0.641
    harris_hpatches-repeat-i: 0.63
    fast_hpatches-repeat-i: 0.576
    shi_hpatches-repeat-i: 0.583
6. Evaluation of the descriptors with homography estimation on HPatches (descriptors_evaluation_on_hpatches.ipynb)
**Changes in viewpoint**
    superpoint_hpatches-des-v: 0.153
    pretrained_mp_hpatches-des-v: 0.231
    sift_hpatches-des-v:  0.254
   orb_hpatches-des-v: 0.122
**Changes in illumination**
    superpoint_hpatches-des-i: 0.849
    pretrained_mp_hpatches-des-i:  0.905
    sift_hpatches-des-i: 0.789
    orb_hpatches-desc-i: 0.491
![TIM图片20190803115214](https://user-images.githubusercontent.com/36001989/62407085-3b631c00-b5e5-11e9-8840-a2bb0d461185.png)

Q1:  Is my step correct?
Q2:  I can't reproduce descriptor evaluation results on HPatches like yours (but detector evaluation results is similar to yours), could you give me some help and advice?

**Comment by yanhuanhuanyy** [comment]
Thanks very much and I am looking forward to your reply.

**Comment by rpautrat** [comment]
Hi,

Your steps seem correct. The only difference is that I exported the pseudo ground truth interest points only twice (I stopped at magic-point_coco-export2). You may need however to tune some parameters to improve the results, like the number of keypoints you keep in the ground truth (parameter 'top_k'). The parameters for the evaluation of the descriptors are also important. For example you need to increase the NMS for bigger images (e.g. 8 for images of size 640x480) and detect more points (e.g. top_k = 1000).

**Comment by yanhuanhuanyy** [comment]
Many thanks for your reply.  I still have some question.
Q1:  About detection,  Is the number of keypoints in the ground truth (parameter 'top_k')  in **magic-point_coco_export.yaml**?  **Magic-point_coco_train.yaml**  file also has parameter 'top_k', do I need to tune the parameter 'top_k' of magic-point_coco_train.yaml when I train the magic-point_coco? If the parameter 'top_k' of magic-point_coco_export.yaml and magic-point_coco_train.yaml need to be tuned, do they need to be tuned to the same value?  300 or 600 or other values?
Q2:  About descriptors, when I evaluate the descriptors, I use the default parameters(nms: 8; 
    top_k: 1000 in  superpoint_hpatches.yaml, classical-descriptors.yaml). But,  when training the superpoint-coco, the nms is  4 and the top_k is 600(superpoint_coco.yaml). They are the default parameters. Do you mean that I need to tune them in superpoint_coco.yaml?
I'm sorry for my verbosity.

**Comment by rpautrat** [comment]
Q1: You don't need to tune the top_k in `magic-point_coco_train.yaml` when training, this one is only important when doing a prediction. The top_k that needs to be tuned during training is the one in `magic-point_coco_export.yaml`. If I remember correctly, I actually commented it in the config file, such that as many points as possible where selected. In that case the real criteria that selected the points was the detection_threshold (which I set to 0.015 I think in `magic-point_coco_export.yaml`). But you can probably tune this detection_threshold to get better performances.

Q2: The nms is not useful during training (since the loss only uses logits which are before the nms), so no need to tune it `superpoint_coco.yaml`. So I guess that you already had the right parameters for the descriptors evaluation.

**Comment by yanhuanhuanyy** [comment]
Thank you for your quick reply!  I'll try it, I still have a small question.
Q3: When I evaluate the repeatability on HPatches, the top_k is 1000 in **magic-point_repeatability.yaml,** the top_k is 600 in **classical-detectors_repeatability.yaml**, they are the default parameters. Do they need to be tuned to the same value for evaluating the repeatability?

**Comment by rpautrat** [comment]
Technically the number of detections should to be the same for a fair comparison of repeatability. In practice I used a top_k of 1000 in `magic-point_repeatability.yaml` to be sure to have a good number of detections, but when doing the evaluation of repeatability, I specified that I wanted to use only the 300 best detections (with the parameter 'keep_k_points' in `compute_repeatability(exper_name, keep_k_points=300, distance_thresh=3, verbose=False`).

**Comment by yanhuanhuanyy** [comment]
Thank you again for your kindness! I got it,  I appreciate your work!
